### DISCUSSION TITLE
Github Package Registry doesn't work as guided

### DISCUSSION THREAD

**Discussion by ralphilius** [vercel/vercel#5016] [Help] [answered]
Hello,

I've tried to setup a private Github Package Registry by following [this guide](https://vercel.com/knowledge/using-private-dependencies-with-vercel) but I keep getting this error:

> error Couldn't find package "@username/xxxx" on the "npm" registry.

I've tried with below setting in my `.npmrc` file:

```
@username=https://npm.pkg.github.com/username
//npm.pkg.github.com/:_authToken=GITHUB_TOKEN
//registry.npmjs.org/
```
I added NPM_RC environment in `vercel.json` file as:

```
{
  "build": {
    "env": {
      "NPM_RC": "@npmrc-github"
    }
  }
}
```

Secrets was added using this command:
```
vc secrets add npmrc-github "$(cat ~/.npmrc)"
```

What did I do wrong in the config?

P.S: I use `yarn` as my package manager

**Comment by styfle** [comment]
`@username=https://npm.pkg.github.com/username` should be `@username:registry=https://npm.pkg.github.com`

See GitHub's instructions here: https://docs.github.com/en/packages/using-github-packages-with-your-projects-ecosystem/configuring-npm-for-use-with-github-packages

**Reply by ralphilius** [reply]
@styfle It didn't work either.

**Reply by styfle** [reply]
Does it work when you run npm install locally?

**Reply by ralphilius** [reply]
@styfle  I have found the issue. I had a local `.npmrc` in my project folder which was not adjusted to your suggestion. After removing it, the deployment was successful. This was a CLI deployment, so the `.npmrc` was uploaded along with other source code. I thought global `NPM_RC`  would override the local setting so I didn't notice it.

**Reply by styfle** [reply]
You can add .npmrc to .vercelignore if you don’t want the CLI to upload it.

**Reply by majdi** [reply]
Hello, I have the same environment and the same issue, I don't understand why :(

```
always-auth=true
@MY_ORGANISATION:registry=https://npm.pkg.github.com
//npm.pkg.github.com/:_authToken=MY_TOKEN
```
Then I just add `vc env add NPM_RC development < ~/.npmrc` (same for production)

And I still have a 401 with my github packages, no `.npmrc` or `.yarnrc` inside the local directory

**Reply by styfle** [reply]
Did you add to preview too? `vc env add NPM_RC preview < ~/.npmrc`

**Reply by majdi** [reply]
Sorry I just saw your answer didn't have the notification :)
I added preview too, I just retry but same issue.

**Reply by ralphilius** [reply]
@majdi Try to add Environment Variables into your `vercel.json` file

**Reply by majdi** [reply]
Well, it just worked when I tried to re-generate my github token...
I used my last token in multiple project and for a long time so I don't know why it worked now but in my despair I just tried too 😅

I tried with vercel last time didn't work well too, but seems that it was from my environment config. Anyway Thanks guys 😄

**Comment by agcty** [comment]
I'm having the same issue.

We're using yarn and to build the project and have a private github npm registry. NPM_RC is correctly set, when using npm, vercel deploys the project without issues but with yarn it always throws an unauthorized error.

![image](https://user-images.githubusercontent.com/12101091/99075183-575fb680-25b9-11eb-9e62-bd198e94f541.png)

**Comment by mknapper1** [comment, marked as answer]
For those coming across this issue I was able to resolve from [this comment](https://github.com/vercel/vercel/discussions/5418#discussioncomment-130437)

You have to make your .npmrc look exactly like this:

```
always-auth=true
@YOUR_USERNAME:registry=https://npm.pkg.github.com/YOUR_USERNAME/
//npm.pkg.github.com/:_authToken=YOUR_TOKEN
//npm.pkg.github.com/YOUR_USERNAME/:_authToken=YOUR_TOKEN
registry=https://registry.npmjs.org
```

**Reply by sivaraj-v** [reply]
Small correction required, 

```
@phanatic:registry=https://npm.pkg.github.com/
//npm.pkg.github.com/:_authToken=TOKEN
@USERNAME:registry=https://npm.pkg.github.com/USERNAME
registry=https://registry.npmjs.org
```

Replace USERNAME & TOKEN

**Comment by tirrth** [comment]
### For Yarn Folks:

What I found from my observation is to remove yarn.lock file from the project, and adding the given snippet below to `.npmrc` or `NPM_RC env variable` works just fine. Nothing else suggested here worked for me though😕

```
<ORGANIZATION>:registry=https://npm.pkg.github.com/
//npm.pkg.github.com/:_authToken=<TOKEN>
```

**UPDATE:** Rather than removing the yarn.lock from repo itself, We can even override the default install command - `yarn install` to `rm -rf yarn.lock && yarn install` or `yarn install --no-lockfile` from Vercel settings, and it will do the work for us

**Comment by Klohto** [comment]
Can you guys seriously update the docs? https://vercel.com/support/articles/using-private-dependencies-with-vercel

Private GitHub Registry requires a different setup for NPM as well as Yarn and the docs are absolutely useless in this.

**Reply by leerob** [reply]
It doesn't seem like there's a clear right answer, based on the responses here. I'm willing to update that guide, but unsure which solution to choose as there's multiple suggested. Maybe I can just link back to here?

**Reply by Klohto** [reply]
Tested out a few solutions and one solution is always consistent (on my end) - using `always-auth=true` and defining registry if your packages reference it (`@ORGNIZATION:registry=https://npm.pkg.github.com`). After that, using the `_authToken` works without issues.

This thread was useful, but the Vercel Build Logs lack any information if NPM_RC env magic is being added or not.

**Reply by Pieparker** [reply]
Thank you, @Klohto. I've passed this on to our docs team!

internal ref: https://github.com/vercel/documentation/issues/118

**Reply by sajadghawami** [reply]
@leerob 

Still not clear to me how to use env vars with vercel and private npm repos...
This [guide](https://vercel.com/guides/using-private-dependencies-with-vercel) says you can add the env NPM_RC, but does not explains what happens after that. 
Will it be automagically used when a `yarn` is used? Does not seem like it.

Also, it misses to explain what will happen locally after `vercel env pull`. Because running `yarn` (or `npm install` respectively) still results in 401.

**Reply by amyegan** [reply]
Hi @sajadghawami! If you set an `NPM_RC` env variable, that will automatically be used for the specified environment. I'm not sure I understand the problem you're describing with the local `vercel env pull`. 

This is a fairly old discussion, so it's a good idea to [start a new discussion](https://github.com/vercel/community/discussions/new) with some more details about your scenario. If you can include a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) or a detailed description of steps to reproduce the problem, then we can work together to find a solution that will work for your project. 🙏
